banderwagon/precomp_multiexp.go

```diff
@@ -1,135 +1,268 @@
+			return fmt.Errorf("serializing 16-bit table for %d-th point: %s", i, err)
 		}
 	}
 
 	return nil
 }
 
+// DeserializePrecomputedLagrange deserializes a PrecomputeLagrange.
+// See SerializePrecomputedLagrange() for the format description.
 func DeserializePrecomputedLagrange(reader io.Reader) (*PrecomputeLagrange, error) {
 	var pcl PrecomputeLagrange
 
 	var numPoints int64
-	err := binary.Read(reader, binary.LittleEndian, &numPoints)
-	if err != nil {
-		return nil, err
+	if err := binary.Read(reader, binary.LittleEndian, &numPoints); err != nil {
+		return nil, fmt.Errorf("deserializing the number of points: %s", err)
 	}
-	pcl.num_points = int(numPoints)
-	pcl.inner = make([]*LagrangeTablePoints, pcl.num_points)
+	pcl.numPoints = int(numPoints)
 
-	var rowLen int64
-	err = binary.Read(reader, binary.LittleEndian, &rowLen)
-	if err != nil {
-		return nil, err
+	// 8-bit table deserialization.
+	var table8BitCount int64
+	if err := binary.Read(reader, binary.LittleEndian, &table8BitCount); err != nil {
+		return nil, fmt.Errorf("deserializing the number of points for 8-bit table: %s", err)
+	}
+	pcl.inner8Bit = make([]*LagrangeTablePoints, table8BitCount)
+	for i := 0; i < int(table8BitCount); i++ {
+		pcl.inner8Bit[i] = &LagrangeTablePoints{}
+		pcl.inner8Bit[i].Deserialize(reader)
 	}
 
-	for i := 0; i < pcl.num_points; i++ {
-
-		pcl.inner[i] = new(LagrangeTablePoints)
-
-		pcl.inner[i].identity.Identity()
-
-		// Deserialize the matrix
-		pcl.inner[i].matrix = make([]bandersnatch.PointAffine, rowLen)
-		for j := int64(0); j < rowLen; j++ {
-			pcl.inner[i].matrix[j] = bandersnatch.ReadUncompressedPoint(reader)
-		}
+	// 16-bit table deserialization.
+	var table16BitCount int64
+	if err := binary.Read(reader, binary.LittleEndian, &table16BitCount); err != nil {
+		return nil, fmt.Errorf("deserializing the number of points for 16-bit table: %s", err)
+	}
+	pcl.inner16Bit = make([]*LagrangeTablePoints, table16BitCount)
+	for i := 0; i < int(table16BitCount); i++ {
+		pcl.inner16Bit[i] = &LagrangeTablePoints{}
+		pcl.inner16Bit[i].Deserialize(reader)
 	}
 
 	return &pcl, nil
 }
 
+// Commit computes the MSM of a set of evaluations.
 func (p *PrecomputeLagrange) Commit(evaluations []fr.Element) Element {
 	var result Element
 	result.Identity()
 
-	for i := 0; i < len(evaluations); i++ {
+	// We use p.inner16Bits for the first 5 group elements.
+	for i := 0; i < len(evaluations) && i < len(p.inner16Bit); i++ {
 		scalar := &evaluations[i]
 
 		if scalar.IsZero() {
 			continue
 		}
 
-		table := p.inner[i]
+		table := p.inner16Bit[i]
 		scalar_bytes_le := scalar.BytesLE()
 
-		for row, byte := range scalar_bytes_le {
-			if byte == 0 {
+		for row := 0; row < 16; row++ {
+			value := uint16(scalar_bytes_le[2*row]) + uint16(scalar_bytes_le[2*row+1])<<8
+			if value == 0 {
 				continue
 			}
-			tp := table.point(row, byte)
+			tp := table.point(row, value)
 			result.AddMixed(&result, *tp)
 		}
 	}
```

Comment:
Note here that we're doing half `AddMixed` compared to the 8-bit window, constructing `value` as the sum of two bytes (since is a 16-bit table).